File: src/main/java/org/example/bankramenserver/domain/transaction/domain/repository/TransactionRepositoryCustom.java
package org.example.bankramenserver.domain.transaction.domain.repository;

import org.example.bankramenserver.domain.transaction.domain.Transaction;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

public interface TransactionRepositoryCustom {

    List<TransactionHistoryRow> findTransactionHistories(
            UUID userId,
            Transaction.TransactionType transactionType,
            LocalDate startDate,
            LocalDate endDate
    );

    List<TransactionHistoryRow> findRecentTransactionHistories(
            UUID userId,
            int limit
    );

    List<UUID> findUserIdsHavingTransactionsBetween(
            LocalDate startDate,
            LocalDate endDate
    );

    boolean existsSameExpenseTransactionBetween(
            UUID userId,
            String description,
            Long amount,
            LocalDate startDate,
            LocalDate endDate
    );
}

File: src/main/kotlin/org/example/bankramenserver/domain/transaction/domain/repository/TransactionRepositoryCustom.kt
package org.example.bankramenserver.domain.transaction.domain.repository

import org.example.bankramenserver.domain.transaction.domain.Transaction
import java.time.LocalDate
import java.util.UUID

interface TransactionRepositoryCustom {
    fun findTransactionHistories(userId: UUID?, transactionType: Transaction.TransactionType?, startDate: LocalDate?, endDate: LocalDate?): List<TransactionHistoryRow>
    fun findRecentTransactionHistories(userId: UUID?, limit: Int): List<TransactionHistoryRow>
    fun findUserIdsHavingTransactionsBetween(startDate: LocalDate?, endDate: LocalDate?): List<UUID>
    fun existsSameExpenseTransactionBetween(userId: UUID?, description: String?, amount: Long?, startDate: LocalDate?, endDate: LocalDate?): Boolean
}
